src/python/pants/core/goals/deploy.py

```diff
@@ -115,51 +128,294 @@ async def publish_process_for_target(
     )
 
 
-async def _all_publish_processes(targets: Iterable[Target]) -> PublishProcesses:
+async def _publish_processes_for_targets(targets: Iterable[Target]) -> PublishProcesses:
     processes_per_target = await MultiGet(
         Get(PublishProcesses, _PublishProcessesForTargetRequest(target)) for target in targets
     )
 
     return PublishProcesses(chain.from_iterable(processes_per_target))
 
 
-async def _invoke_process(
-    console: Console,
-    process: InteractiveProcess | None,
-    *,
-    names: Iterable[str],
-    success_status: str,
-    description: str | None = None,
-) -> tuple[int, tuple[str, ...]]:
-    results = []
+@dataclass(frozen=True)
+class _DeployTargetRequest(EngineAwareParameter):
+    target: CoarsenedTarget
+    dependent: Address | None = None
 
-    if not process:
-        sigil = console.sigil_skipped()
-        status = "skipped"
-        if description:
-            status += f" {description}"
-        for name in names:
-            results.append(f"{sigil} {name} {status}.")
-        return 0, tuple(results)
+    def debug_hint(self) -> str:
+        result = self.target.representative.address.spec
```

Comment:
I think that your use of `CoarsenedTarget` for this seems fine: it is designed to be used recursively like this.

But the reason a `CoarsenedTarget` contains multiple targets is that it supports handling cycles: a cycle in the graph will result in a `CT` containing multiple targets. From a deployment perspective, you should likely treat that as an error (unless a deployment backend supports deploying multiple targets "at once", the same way a compiler supports compiling multiple files at once).

You should likely add some validation to `_DeployTargetRequest` that the `CT` it is given only contains a single target (and fail with a "Cannot deploy a graph that contains a cycle: `A -> B -> C`": that would make your use of `CT.representative` accurate.

Reply:
thanks, that's a good point, I'll add a check to verify that each CT only has one member.